### PR TITLE
[generation] Support cache-cropping methods

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Adds the requirements to solve #35168 

[EDITED after the discussion ending on [this comment](https://github.com/huggingface/transformers/pull/35591#issuecomment-2586726433)]

New dynamic cache compression methods may crop the cache differently at each layer. Doing so breaks our implicit assumption regarding the shape of the dynamic cache, where all layers have the same length. As we can see in the issue tagged above, we obtain an exception if we decide to do so.

We can enable it with small code changes in downstream libraries if we prepare the causal mask based on the maximum sequence length seen in all layers in the cache. A downstream library would then have to implement a custom attention forward pass, to *left-crop* the attention mask accordingly (e.g. using the length of the `key`)

_____________________________
The following tests were run, with no regressions compared to `main`:
- `RUN_SLOW=1 py.test tests/utils/test_cache_utils.py -vv`
- `RUN_SLOW=1 py.test tests/models/llama/test_modeling_llama.py -vv`

### CODE REVIEW

File: src/transformers/models/llama/modeling_llama.py [hunk 1]
```diff
@@ -660,7 +668,7 @@ def _update_causal_mask(
             target_length = (
                 attention_mask.shape[-1]
                 if isinstance(attention_mask, torch.Tensor)
-                else past_seen_tokens + sequence_length + 1
+                else past_seen_tokens + sequence_length
```
Comment (gante) [inline src/transformers/models/llama/modeling_llama.py:671]: this should go in regardless of whether this draft is accepted -- we are creating a too large causal mask, that we then right-crop in L224 (`causal_mask = attention_mask[:, :, :, : key_states.shape[-2]]`)

Comment (zucchini-nlp) [reply src/transformers/models/llama/modeling_llama.py:671]: indeed, not sure if there was a particular reason for `+1` initially though

**Review by zucchini-nlp — COMMENTED** [review]
As it solves the issue for https://github.com/huggingface/transformers/issues/35168 with no major code changes from KVPress side, LGTM

My only concern is that it is slightly breaking, if users were somehow exploiting this right-side slicing feature before

**Comment by gante** [comment]
@zucchini-nlp have a look at this draft :)

**Comment by gante** [comment]
@ArthurZucker, this draft PR makes a small modeling change to enable an advanced cache feature -- see the PR header for full details.

If you agree with these changes, I will propagate them to other models before requesting a final review :)

**Review by ArthurZucker — COMMENTED** [review]
Hey! Sorry but I don't think it makes sense to do this! 
It's a very niche case, and is not transparent for people who just want to see how llama works!

**Comment by gante** [comment]
@ArthurZucker makes sense. In fact, downstream applications can implement their own custom attention implementation and crop the mask as they wish there.

We would be able to unblock this cool use case in other libs if we merge the non-modeling change, i.e. make the `get_seq_length(layer=None)` call return the longest sequence length for all layers, as opposed to the sequence length for layer 0 (current behavior). The two are equivalent except on this advanced use case. This change looks okay to you, correct?

File: src/transformers/models/aria/modeling_aria.py [hunk 2]
```diff
@@ -1037,7 +1037,7 @@ def _update_causal_mask(
             target_length = (
                 attention_mask.shape[-1]
                 if isinstance(attention_mask, torch.Tensor)
-                else past_seen_tokens + sequence_length + 1
+                else past_seen_tokens + sequence_length
```
Comment (gante) [inline src/transformers/models/aria/modeling_aria.py:1040]: This comment applies to all models: the `+1` not needed, we are discarding its corresponding data in the attn forward pass (e.g. [here](https://github.com/huggingface/transformers/blob/2fa876d2d824123b80ced9d689f75a153731769b/src/transformers/models/llama/modeling_llama.py#L220))

By removing it, we should expect a minimal improvement wrt memory and inference time. We also ensure that the causal mask has exactly the length of the (cached) input length, which is a requirement for the downstream use case this PR wants to enable.

**Comment by zucchini-nlp** [comment]
Btw, I am working with Mllama now and remembered it has weird cache where cross attention layers have a very large length. Can we make sure mllama generation tests dont fail after this change?

**Comment by gante** [comment]
@zucchini-nlp correct, it does change. Can you think of more models that a) have cross attn cache AND b) don't use `EncoderDecoderCache`, like mllama?

btw, `RUN_SLOW=1 py.test tests/models/mllama/test_modeling_mllama.py` results in 11 failures on both `main` and this branch 😬

**Comment by zucchini-nlp** [comment]
That is too much for the `main` branch, afaik mllama will not be able to run the slow tests because we are in EU and the repo is gated 😢  

Currently I am removing skip from many generation tests that checked `hasattr(config, "use_cache")` otherwise all of them are skipped in multimodal models. In mlllama case it also skips all tests that need caching I guess. From code it seems like the `get_seq_length()` is used only when preparing `cache_position` if `None` and when preparing the attention mask. So I think the model might fail whenever there is no mask to infer length or when continuing generation from "initial prompt"


For other models, in VLM side that's the only one and maybe ImageGPT is a bit peculiar. From LLM side all I can think is the Hybrid Cache models where sliding window length is fixed and can be lower/higher than static layers

**Review by ArthurZucker — COMMENTED** [review]
not sure where this stands at, but changes look better now that they are super minimal!

**Comment by gante** [comment]
@ArthurZucker the PR is ready :) 

The changes were originally made to help an external repo, but they are good (IMO) general fixes -- when we request the length of the cache, let's look at all layers instead of just looking at layer 0.
